packages/circuits/src/shuffle.circom

```diff
@@ -0,0 +1,35 @@
+pragma circom 2.0.0;
+
+include "./merkle.circom";
+
+template Shuffle(levels) {
+
+    signal input initialDeck[2**levels];
+    signal input finalDeck[2**levels];
+    signal input deckPredicate[2**levels];
+    signal input deckRoot;
+
+    var maxDeckSize = 2**levels;
+
+    var deckSum;
+    for (var i = 0; i < maxDeckSize; i++) {
+        deckSum += initialDeck[i] * (maxDeckSize**i);
+    }
+
+    signal tempSum[2**levels]; // we need to use intermediate signal to prevent non quadratic error
+    for (var i = 0; i < maxDeckSize; i++) {
+        if (i == 0) {
+            tempSum[i] <== finalDeck[i];
+        } else {
+            tempSum[i] <== tempSum[i-1] + (finalDeck[i] * deckPredicate[i]);
```

Comment:
So I'm thinking ... is this even secure? 😅 

To prove the shuffle `[x, y, z] → [x, y, z]`, you would prove

```x * 1 + y * 64 + z * 64^2 ==
x * 1 + z * 64 ^2 + y * 64
```

But then, you could prove the (incorrect) shuffle `[x, y, z] → [x, y, w]` by proving
```
x * 1 + y * 64 + z * 64^2 ==
x * 1 + y * 64 + w * (z * 64^2 / w)
```

I think the solution is simply to constrain the `deckPredicate` array by computing `sum(deckSize ** i)` and ensuring the sum over `deckPredicate` is equal to that.

Googled a bit and found this: https://zkresear.ch/t/random-permutation-circuit/74
It uses a random value (`hash`) to shuffle the deck. We could allow just any value to be passed for that. Might be a lot more efficient than what we have right now (4.3k constraints says Jordi).

Haven't studied the circuit yet but I intend to! Wonder if that could also be used to reduce the complexity of the original hand proof.

Reply:
Oooh this is a very interesting observation! I have constrained the deck predicate for now - thanks for sharing the permutation circuit, checking it out..